File: sim/spy.go
package sim

import (
	"fmt"
)

type Spy struct {
	count int
}

func (t *Spy) Start() uint16 {
	return 0xA000
}

func (t *Spy) Length() uint32 {
	return 2
}

func (t *Spy) Read(addr uint16) uint8 {
	panic("not implemented")
}

func (t *Spy) Write(addr uint16, val uint8) {
	t.count++
	if addr == 0xA000 {
		fmt.Printf("[SPY] [W] [DISK_R] COUNT %d VALUE %04x %04d(DEC) %s\n", t.count, val, val, string(val))
		return
	}
	if addr == 0xA001 {
		fmt.Printf("[SPY] [W] [DISK_W] COUNT %d VALUE %04x %04d(DEC) %s\n", t.count, val, val, string(val))
	}
}

File: sim/src/spy.c
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include "devices.h"

// Implement the Start method
uint16_t Spy_Start() {
    return 0x8010;
}

// Implement the Length method
uint32_t Spy_Length() {
    return 5;
}

// Implement the Read method
uint8_t Spy_Read(uint16_t addr, Spy *spy) {
    printf("[SPY] reading addr: %04x", addr);
    return 0;
}

// Implement the Write method
void Spy_Write(uint16_t addr, uint8_t val, Spy *spy) {
    printf("[SPY] writing addr: %04x - value %d (dec) %c (s) \n", addr, val, (char)val);
}


// Function to create a Ram object and initialize the Space interface
Spy* CreateSpy() {
    Spy* spy = malloc(sizeof(Spy));

    return spy;
}